Improve execute() Cypher error when node property in a param of CREATE is null

This:

``` groovy
# literalProps.t is null
final literalProps = [t: literal.literalDatatypeURI, v: literalValue]
final merge = """
MERGE (subj:Resource {href: {subjectHref}})
CREATE (subj) -[:$relName {relProps}]-> (lit:Literal {literalProps})
"""
final params = [subjectHref: subjectHref,
    relProps   : relProps, literalProps: literalProps] as Map<String, Object>
exec.execute(merge, params)
```

throws:

```
java.lang.RuntimeException: Cannot execute: «
MERGE (subj:Resource {href: {subjectHref}})
CREATE (subj) -[:yago_isPreferredMeaningOf {relProps}]-> (lit:Literal {literalProps})
» params: [subjectHref:yago:Dr._Kuno_Struck_House, relProps:[f:yago:id_893bmq_nyg_18f46ch], literalProps:[t:null, v:Dr. Kuno Struck House, l:eng]]
    at id.ac.itb.lumen.persistence.LumenPersistenceImportApp$_importFile_closure1$_closure2.doCall(LumenPersistenceImportApp.groovy:159)
    at id.ac.itb.lumen.persistence.LumenPersistenceImportApp$_importFile_closure1$_closure2.call(LumenPersistenceImportApp.groovy)
    at org.codehaus.groovy.runtime.IOGroovyMethods.withCloseable(IOGroovyMethods.java:1618)
    at id.ac.itb.lumen.persistence.LumenPersistenceImportApp$_importFile_closure1.doCall(LumenPersistenceImportApp.groovy:70)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
    at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:324)
    at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:292)
    at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1017)
    at groovy.lang.Closure.call(Closure.java:423)
    at groovy.lang.Closure.call(Closure.java:439)
    at org.codehaus.groovy.runtime.IOGroovyMethods.withReader(IOGroovyMethods.java:1179)
    at id.ac.itb.lumen.persistence.LumenPersistenceImportApp.importFile(LumenPersistenceImportApp.groovy:67)
    at id.ac.itb.lumen.persistence.LumenPersistenceImportApp.run(LumenPersistenceImportApp.groovy:234)
    at org.springframework.boot.SpringApplication.runCommandLineRunners(SpringApplication.java:677)
    ... 9 common frames omitted
Caused by: scala.MatchError: (t,null) (of class scala.Tuple2)
    at org.neo4j.cypher.internal.compiler.v2_1.mutation.CreateNode$$anonfun$org$$$$cc5e771a9d2ffdf71ac346e6f5f9cb3$$$$e$$fromAnyToLiteral$1$1.apply(CreateNode.scala:41)
    at org.neo4j.cypher.internal.compiler.v2_1.mutation.CreateNode$$anonfun$org$$$$cc5e771a9d2ffdf71ac346e6f5f9cb3$$$$e$$fromAnyToLiteral$1$1.apply(CreateNode.scala:41)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
    at scala.collection.Iterator$class.foreach(Iterator.scala:727)
    at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
    at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
    at scala.collection.AbstractTraversable.map(Traversable.scala:105)
    at org.neo4j.cypher.internal.compiler.v2_1.mutation.CreateNode.org$neo4j$cypher$internal$compiler$v2_1$mutation$CreateNode$$fromAnyToLiteral$1(CreateNode.scala:41)
    at org.neo4j.cypher.internal.compiler.v2_1.mutation.CreateNode$$anonfun$exec$1.apply(CreateNode.scala:78)
    at org.neo4j.cypher.internal.compiler.v2_1.mutation.CreateNode$$anonfun$exec$1.apply(CreateNode.scala:74)
    at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
    at org.neo4j.cypher.internal.helpers.CollectionSupport$$anon$1.next(CollectionSupport.scala:52)
    at scala.collection.Iterator$$anon$13.next(Iterator.scala:372)
    at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
    at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
    at org.neo4j.cypher.internal.compiler.v2_1.pipes.EmptyResultPipe.internalCreateResults(EmptyResultPipe.scala:29)
    at org.neo4j.cypher.internal.compiler.v2_1.pipes.PipeWithSource.createResults(Pipe.scala:105)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1$$anonfun$apply$2.apply(ExecutionPlanBuilder.scala:120)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1$$anonfun$apply$2.apply(ExecutionPlanBuilder.scala:119)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionWorkflowBuilder.runWithQueryState(ExecutionPlanBuilder.scala:168)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:118)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:103)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionPlanBuilder$$anon$1.execute(ExecutionPlanBuilder.scala:68)
    at org.neo4j.cypher.internal.compiler.v2_1.executionplan.ExecutionPlanBuilder$$anon$1.execute(ExecutionPlanBuilder.scala:67)
    at org.neo4j.cypher.internal.ExecutionPlanWrapperForV2_1.execute(CypherCompiler.scala:159)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:76)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:71)
    at org.neo4j.cypher.javacompat.ExecutionEngine.execute(ExecutionEngine.java:84)
    at id.ac.itb.lumen.persistence.LumenPersistenceImportApp$_importFile_closure1$_closure2.doCall(LumenPersistenceImportApp.groovy:156)
    ... 26 common frames omitted
```

It should be handled by mentioning something like: `Invalid value 'null' for parameter 'literalProps.t'. Cannot execute Cypher query: <query> params: <params>`

--- a/community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateAcceptanceTest.scala
+++ b/community/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateAcceptanceTest.scala
@@ -25,6 +25,43 @@ import org.neo4j.cypher.{ExecutionEngineFunSuite, NewPlannerTestSupport, QuerySt
 class CreateAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTestSupport with NewPlannerTestSupport
   with CreateTempFileTestSupport {
 
+  test("handle null value in property map from parameter for create node") {
+    val query = "CREATE (a {props}) RETURN a.foo, a.bar"
+
+    val result = updateWithBothPlanners(query, "props" -> Map("foo" -> null, "bar" -> "baz"))
+
+    result.toSet should equal(Set(Map("a.foo" -> null, "a.bar" -> "baz")))
+    assertStats(result, nodesCreated = 1, propertiesWritten = 1)
+  }
+
+  test("handle null value in property map from parameter for create node with SET") {
+    createNode(("foo", 42), ("bar", "fu"))
+    val query = "MATCH (a) SET a = {props} RETURN a.foo, a.bar"
+
+    val result = updateWithBothPlanners(query, "props" -> Map("foo" -> null, "bar" -> "baz"))
+
+    result.toSet should equal(Set(Map("a.foo" -> null, "a.bar" -> "baz")))
+    assertStats(result, propertiesWritten = 2)
+  }
+
+  test("handle null value in property map from parameter for create relationship") {
+    val query = "CREATE (a)-[r:REL {props}]->() RETURN r.foo, r.bar"
+
+    val result = updateWithBothPlanners(query, "props" -> Map("foo" -> null, "bar" -> "baz"))
+
+    result.toSet should equal(Set(Map("r.foo" -> null, "r.bar" -> "baz")))
+    assertStats(result, nodesCreated = 2, relationshipsCreated = 1, propertiesWritten = 1)
+  }
+
+  test("handle null value in property map from parameter") {
+    val query = "CREATE (a {props})-[r:REL {props}]->() RETURN a.foo, a.bar, r.foo, r.bar"
+
+    val result = updateWithBothPlanners(query, "props" -> Map("foo" -> null, "bar" -> "baz"))
+
+    result.toSet should equal(Set(Map("a.foo" -> null, "a.bar" -> "baz", "r.foo" -> null, "r.bar" -> "baz")))
+    assertStats(result, nodesCreated = 2, relationshipsCreated = 1, propertiesWritten = 2)
+  }
+
   //Not TCK material
   test("should have bound node recognized after projection with WITH + LOAD CSV") {
     val url = createCSVTempFileURL( writer => writer.println("Foo") )

--- a/community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/mutation/CreateNode.scala
+++ b/community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/mutation/CreateNode.scala
@@ -20,7 +20,7 @@
 package org.neo4j.cypher.internal.compiler.v3_1.mutation
 
 import org.neo4j.cypher.internal.compiler.v3_1._
-import org.neo4j.cypher.internal.compiler.v3_1.commands.expressions.{Expression, Literal}
+import org.neo4j.cypher.internal.compiler.v3_1.commands.expressions.{Expression, Literal, Null}
 import org.neo4j.cypher.internal.compiler.v3_1.commands.values.KeyToken
 import org.neo4j.cypher.internal.compiler.v3_1.executionplan.{Effects, _}
 import org.neo4j.cypher.internal.compiler.v3_1.helpers.ListSupport
@@ -44,6 +44,7 @@ case class CreateNode(key: String, properties: Map[String, Expression], labels: 
   def exec(context: ExecutionContext, state: QueryState): Iterator[ExecutionContext] = {
     def fromAnyToLiteral(x: Map[String, Any]): Map[String, Expression] = x.map {
       case (k, v:Any) => k -> Literal(v)
+      case (k, null) => k -> Null()
     }
 
     def createNodeWithPropertiesAndLabels(props: Map[String, Expression]): ExecutionContext = {

--- a/community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/mutation/GraphElementPropertyFunctions.scala
+++ b/community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/mutation/GraphElementPropertyFunctions.scala
@@ -66,11 +66,13 @@ trait GraphElementPropertyFunctions extends ListSupport {
 
     pc match {
       case n: Node => map.foreach {
+        case (_, null) =>
         case (key, value) =>
           state.query.nodeOps.setProperty(n.getId, state.query.getOrCreatePropertyKeyId(key), makeValueNeoSafe(value))
       }
 
       case r: Relationship => map.foreach {
+        case (_, null) =>
         case (key, value) =>
           state.query.relationshipOps.setProperty(r.getId, state.query.getOrCreatePropertyKeyId(key), makeValueNeoSafe(value))
       }

